src/jsonrpc/mod.rs

```diff
@@ -0,0 +1,134 @@
+use anyhow::{anyhow, Result};
+use async_channel::{Receiver, Sender};
+use async_trait::async_trait;
+use futures_util::{SinkExt, StreamExt};
+use reqwest::Client;
+use reqwest::header::{CONTENT_TYPE, HeaderMap, HeaderValue};
+use serde_json::json;
+use serde_json::Value;
+use tokio::net::TcpStream;
+use tokio::spawn;
+use tokio_tungstenite::{connect_async, WebSocketStream};
+use tokio_tungstenite::MaybeTlsStream;
+use tokio_tungstenite::tungstenite::Message;
+use url::Url;
+
+#[cfg(test)]
+mod tests;
+
+/// A convenience constant that represents empty params in a JSON-RPC request.
+pub const NO_PARAMS: Value = json!([]);
+
+/// A simple async JSON-RPC client that can send one-shot request via HTTP/HTTPS
+/// and subscribe to push-based notifications via Websockets. The returned
+/// results are of type [`Value`] from the [`serde_json`] crate.
+#[async_trait]
+pub trait JsonRpcClient {
+    /// Sends a JSON-RPC request with `method` and `params` via HTTP/HTTPS.
+    async fn request(&self, method: &str, params: Value) -> Result<Value>;
+
+    /// Subscribes to notifications via a Websocket. This returns a [`Receiver`]
+    /// channel that is used to receive the messages sent by the server.
+    async fn subscribe(&self, method: &str) -> Result<Receiver<Value>>;
+}
+
+/// The implementation of [`JsonRpcClient`].
+pub struct JsonRpcClientImpl {
+    http_client: Client,
+    url: Url,
+}
+
+impl JsonRpcClientImpl {
+    /// Creates a client that sends all requests to `url`.
+    pub fn new(url: Url) -> Self {
+        Self { http_client: Client::default(), url }
+    }
+}
+
+#[async_trait]
+impl JsonRpcClient for JsonRpcClientImpl {
+    async fn request(&self, method: &str, params: Value) -> Result<Value> {
+        let request_body = build_jsonrpc_request(method, params)?;
+        let response = self.http_client.post(self.url.as_str())
+            .headers(HeaderMap::from_iter(
+                [(CONTENT_TYPE, HeaderValue::from_static("application/json"))]))
```

Comment:
We will need to access operations from Lotus with an `admin/write` role, so we will have to attach the authentication token as part of the header in the `Bearer` field. See https://lotus.filecoin.io/reference/basics/api-access/#locally-hosted-node

Reply:
Added support for the bearer token